src/crystal/event_loop/io_uring.cr

```diff
@@ -0,0 +1,483 @@
+    ts.value.tv_nsec = typeof(ts.value.tv_nsec).new!(timeout.@nanoseconds)
+
+    sqe.value.opcode = LibC::IORING_OP_TIMEOUT
+    sqe.value.user_data = event.address.to_u64!
+    sqe.value.addr = ts.address.to_u64!
+    sqe.value.len = 1
+    trace(sqe)
+
+    @ring.submit
+  end
+
+  def delete_timer(event : Event*) : Nil
+    sqe = @ring.next_sqe
+    sqe.value.opcode = LibC::IORING_OP_TIMEOUT_REMOVE
+    sqe.value.flags = LibC::IOSQE_CQE_SKIP_SUCCESS
+    sqe.value.addr = event.address.to_u64!
+    trace(sqe)
+
+    @ring.submit
+  end
+
+  # fiber interface, see Crystal::EventLoop
+
+  def sleep(duration : Time::Span) : Nil
+    res = async_timeout(:sleep, duration)
+    # assert(res == -ETIME)
+  end
+
+  def create_timeout_event(fiber : Fiber) : FiberEvent
+    FiberEvent.new(:select_timeout, fiber)
+  end
+
+  # file descriptor interface, see Crystal::EventLoop::FileDescriptor
+
+  def read(file_descriptor : System::FileDescriptor, slice : Bytes) : Int32
+    async_rw(LibC::IORING_OP_READ, file_descriptor, slice, file_descriptor.@read_timeout) do |errno|
+      case errno
+      when Errno::ECANCELED
+        raise IO::TimeoutError.new("Read timed out")
+      when Errno::EBADF
+        raise IO::Error.new("File not open for reading", target: file_descriptor)
+      else
+        raise IO::Error.from_os_error("read", errno, target: file_descriptor)
+      end
+    end
+  end
+
+  def wait_readable(file_descriptor : System::FileDescriptor) : Nil
+    async_poll(file_descriptor, LibC::POLLIN | LibC::POLLRDHUP, file_descriptor.@read_timeout) { "Read timed out" }
+  end
+
+  def write(file_descriptor : System::FileDescriptor, slice : Bytes) : Int32
+    async_rw(LibC::IORING_OP_WRITE, file_descriptor, slice, file_descriptor.@write_timeout) do |errno|
+      case errno
+      when Errno::ECANCELED
+        raise IO::TimeoutError.new("Write timed out")
+      when Errno::EBADF
+        raise IO::Error.new("File not open for writing", target: file_descriptor)
+      else
+        raise IO::Error.from_os_error("write", errno, target: file_descriptor)
+      end
+    end
+  end
+
+  def wait_writable(file_descriptor : System::FileDescriptor) : Nil
+    async_poll(file_descriptor, LibC::POLLOUT, file_descriptor.@write_timeout) { "Write timed out" }
+  end
+
+  def reopened(file_descriptor : System::FileDescriptor) : Nil
+    # TODO: do we need to cancel pending operations?
+  end
+
+  def close(file_descriptor : System::FileDescriptor) : Nil
+    # sync with `FileDescriptor#file_descriptor_close`: prevent actual close
+    fd = file_descriptor.@volatile_fd.swap(-1, :relaxed)
+    return if fd == -1
+
+    async_close(fd) do |sqe|
+      # one thread closing a fd won't interrupt reads or writes happening in
+      # other threads, for example a blocked read on a fifo will keep blocking,
+      # while close would have finished and closed the fd; we thus explicitly
+      # cancel any pending operations on the fd before we try to close
```

Comment:
The [close(2)](https://man7.org/linux/man-pages/man2/close.2.html#NOTES) manpage explicitly states that some systems interrupt any blocking `read` or `write` but the Linux behavior is to :see_no_evil: